server/src/main/java/io/kroki/server/service/Plantuml.java

```diff
@@ -177,6 +226,43 @@ private static void ignoreInclude(String line, StringBuilder sb) {
+        && !include.startsWith("http://") && !include.startsWith("https://") // no URLs
+        // no path walking
+        && !include.startsWith("../") && !include.contains("/../") && !include.endsWith("/..")
+        && !include.startsWith("..\\") && !include.contains("\\..\\") && !include.endsWith("\\..")
+        && !include.contains("/..\\") && !include.contains("\\../")) {
+        // PlantUML's safety checking will suffice as it will only allow files directly in the include path
+
+        // Note: we are relying on the PlantUML include checking algorithm:
+        // * See ImportedFiles#executeInclude and ImportedFiles#executeIncludesub
+        // * !importsub does not work with URLs or the standard library
+        // * Imports that start with `http://` or `https://` are resolved from the URL, that is generally
+        //   unsafe unless you can trust the source of the URL, thus we do not permit here.
+        // * Imports wrapped in `<` and `>` are resolved only from the standard library and thus should be
+        //   a fixed set guarded by `/stdlib/${name}-abx.repx` resources on the classpath, in any case we
+        //   ignore stdlib import validation here
+        // * Absolute imports could refer to any file, so we exclude those, though FileWithSuffix#fileOk()
+        //   should restrict accessible files to only those "approved"
+        // * We only auto-permit "search path" includes that do not attempt search path escaping with `..`
+        // * The import or importsub that we auto-permit will result in a call to ImportedFiles.getFile(...)
+        // * That will iterate the "plantuml.include.path" (note PlantUML is not well designed for embedding
+        //   on this property as the ImportedFiles#INCLUDE_PATH is initialized once on classloading)
+        // * At this point, if the INCLUDE_PATH included say `/example` and the include was `foo/bar.puml`
+        //   and there is a file `/example/foo/bar/puml` then that would stop the search with an `AFileRegular`
+        //   instance. HOWEVER, the return value of ImportedFiles#getFile is guarded by ImportedFiles#isAllowed
+        // * ImportedFiles#isAllowed will only permit AFile instances with the AFile#getSystemFolder() being
+        //   contained in ImportedFiles#INCLUDE_PATH so as AFileRegular#getSystemFolder() always returned the
+        //   parent folder of the file and include of `foo/bar.puml` found on an include path of `/example`
+        //   resolving to `/example/foo/bar.puml` will have a system folder of `/example/foo` which is not
+        //   one of the folders whitelisted in the include path and thus the include will be resolved as not-found
+        //
+        // Effectively only imports that are immediate children of the folders listed in "plantuml.include.path"
+        // are eligible for inlcude (unless OptionFlags.ALLOW_INCLUDE has been set to true)
+        sb.append(line).append("\n");
```

Comment:
I really appreciate the time and effort you put into this clear and precise analysis of the PlantUML include checking algorithm :100: 

If I understand correctly, even if we do allow path walking, the PlantUML include algorithm should not resolve the include if the path is outside of the include path right? So if my include path is `/example` and my include is `../etc/passwd`, the system folder of the file will be `/etc` and `/etc` is not part of the include path `/example` so the include will be resolved as not-found. Am I correct?
I'm not saying that we should remove the path walking checks (as we want to disallow them as soon as possible) but I want to make sure that I understand how it's working.

Reply:
correct, but please verify my claim rather than take it at face value. I believe it to be correct, but you are the maintainer, so "trust but verify" especially as this as security implications